### DISCUSSION TITLE
ESM1v fine-tuning details (spiked finetuning)

### DISCUSSION THREAD

**Discussion by Stefan-Ghivi** [facebookresearch/esm#506] [Q&A]
Hello, 

I've recently discovered your ESM1v paper for variant prediction. Congrats on all the work. 
I was able to reproduce and use your zero-shot  approach with the pre-trained model you provided. 
Now I'm trying to fine-tune the model following the spiked fine-tuning procedure described in the paper, but I have some question and I was hoping maybe you could help.

1. In #153 there is a link for the MSA generation code. It is somewhat unclear to me how these MSAs are used in finetuning. As per evo-tuning should I use all the sequences from MSA with a masked language model training objective ? If so, how should I process the gaps ("-" characters) from the aligned sequences ? Do I simply remove them? Do I use a mask/ pad token for them? 

2. How many sequences should an MSA contain? In the paper there is a mention for 384 sequences, but I believe that is for the MSA Transformer inference. What about the ESM1v fine tuning?

3. Referring to the spiked fine-tuning process: 0.01 out of 500k tokens = 5k tokens should be a)MSA tokens or b)pre-training data tokens? How are the pre-training sequences selected for this? Also, there a mention : " In this setting, we train on the entire MSA, including
non-consensus positions". How should I process the non-consensus positions? mask/ pad them? 

4. If I understand correctly, you fine-tune with a batch size of 500k tokens. Max sequence length supported by ESM1v is 1024 tokens. So your batch size is roughly 500 sequences? How do you manage such a large batch size? Some kind of ddp + gradient accumulation strategy?

I hope you guys can shed some light on this,
Thanks for reading!

**Comment by tomsercu** [comment]
1. was actually mentioned as an experiment:
> We also assessed label smoothing and replacing the gap token with a mask token or a pad token finding no significant impact; for simplicity, we omit label smoothing and use the mask token for future experiments.
2. the whole MSA
3. Spiked fine-tuning: 1% msa tokens, 99% other sequences (=strong regularization). Non consensus positions are just uppercased and treated as normal characters
4.  Yes DDP across many gpus, and accumulating gradients from several forward&backward passes before doing an optimizer step (which reduces communication as well, good for GPU utilziation). See https://fairseq.readthedocs.io/en/latest/getting_started.html#large-mini-batch-training-with-delayed-updates

Hope this helps!

**Comment by Stefan-Ghivi** [comment]
Hi @tomsercu thanks for your answer, it was very helpful in clarifying things for me.
I've started the fine-tuning process and now 2 more questions bug me:

5. When you finetuned on the 10 validation datasets, you trained on all of them at once? Or did you at any point trained the model separately for each of the datasets ? I'm asking because I'm trying to finetune the model for a particular dataset so in that sense I would like to make the model to overfit on that data. 

6. If I have a long sequence I want to fine-tune on (let's say 2k aa long), when computing the MSA, should I use a)the entire sequence (all 2k aa); b) a crop of 1024 aa; c) only the area of interest where the mutational experiments occurred (let's say ~200aa) ?

I guess these are more like problems that I would appreciate your advice on rather than questions.
Thanks for reading!

**Comment by tomsercu** [comment]
5. finetuning per variant prediction task
6. long sequences are a bit tricky. It's reasonable to do random crops so over time you see the whole sequence length, or a domain if you know that is conserved and meaningful in itself.

**Comment by Stefan-Ghivi** [comment]
Hi @tomsercu, could you please clarify the finetuning procedure for me? 

So the 9 datasets from validation set are only used for "modeling purposes" and a model is trained for each dataset? That means that for full 41 datasets few-shot metric I need 41 models? I was under the impression that you use the 9 datasets to train a single model that is afterwards used for few-shoting the test datasets. 

Thanks!
